util/winutil/restartmgr_windows.go

```diff
@@ -0,0 +1,830 @@
+)
+
+// Implementation note: the code in this file will be invoked from within
+// MSI custom actions, so please try to return windows.Errno error codes
+// whenever possible; this makes the action return more accurate errors to
+// the installer engine.
+
+const (
+	_RESTART_NO_CRASH  = 1
+	_RESTART_NO_HANG   = 2
+	_RESTART_NO_PATCH  = 4
+	_RESTART_NO_REBOOT = 8
+)
+
+func registerForRestart(opts RegisterForRestartOpts) error {
+	var flags uint32
+
+	if !opts.RestartOnCrash {
+		flags |= _RESTART_NO_CRASH
+	}
+	if !opts.RestartOnHang {
+		flags |= _RESTART_NO_HANG
+	}
+	if !opts.RestartOnUpgrade {
+		flags |= _RESTART_NO_PATCH
+	}
+	if !opts.RestartOnReboot {
+		flags |= _RESTART_NO_REBOOT
+	}
+
+	var cmdLine *uint16
+	if opts.UseCmdLineArgs {
+		if len(opts.CmdLineArgs) == 0 {
+			// re-use our current args, excluding the exe name itself
+			opts.CmdLineArgs = os.Args[1:]
+		}
+
+		var b strings.Builder
+		for _, arg := range opts.CmdLineArgs {
+			if b.Len() > 0 {
+				b.WriteByte(' ')
+			}
+			b.WriteString(windows.EscapeArg(arg))
+		}
+
+		if b.Len() > 0 {
+			var err error
+			cmdLine, err = windows.UTF16PtrFromString(b.String())
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	hr := registerApplicationRestart(cmdLine, flags)
+	if e := wingoes.ErrorFromHRESULT(hr); e.Failed() {
+		return e
+	}
+
+	return nil
+}
+
+type _RMHANDLE uint32
+
+type _RM_UNIQUE_PROCESS struct {
+	PID              uint32
+	ProcessStartTime windows.Filetime
+}
+
+type _RM_APP_TYPE int32
+
+const (
+	_RmUnknownApp  _RM_APP_TYPE = 0
+	_RmMainWindow  _RM_APP_TYPE = 1
+	_RmOtherWindow _RM_APP_TYPE = 2
+	_RmService     _RM_APP_TYPE = 3
+	_RmExplorer    _RM_APP_TYPE = 4
+	_RmConsole     _RM_APP_TYPE = 5
+	_RmCritical    _RM_APP_TYPE = 1000
+)
+
+type _RM_APP_STATUS uint32
+
+const (
+	_RmStatusUnknown        _RM_APP_STATUS = 0x0
+	_RmStatusRunning        _RM_APP_STATUS = 0x1
+	_RmStatusStopped        _RM_APP_STATUS = 0x2
+	_RmStatusStoppedOther   _RM_APP_STATUS = 0x4
+	_RmStatusRestarted      _RM_APP_STATUS = 0x8
+	_RmStatusErrorOnStop    _RM_APP_STATUS = 0x10
+	_RmStatusErrorOnRestart _RM_APP_STATUS = 0x20
+	_RmStatusShutdownMasked _RM_APP_STATUS = 0x40
+	_RmStatusRestartMasked  _RM_APP_STATUS = 0x80
+)
+
+type _RM_PROCESS_INFO struct {
+	Process          _RM_UNIQUE_PROCESS
+	AppName          [256]uint16
+	ServiceShortName [64]uint16
+	AppType          _RM_APP_TYPE
+	AppStatus        _RM_APP_STATUS
+	TSSessionID      uint32
+	Restartable      int32 // Win32 BOOL
+}
+
+// RestartManagerSession represents an open Restart Manager session.
+type RestartManagerSession interface {
+	io.Closer
+	// AddPaths adds the fully-qualified paths in fqPaths to the set of binaries
+	// that will be monitored by this restart manager session.
+	AddPaths(fqPaths []string) error
+	// AffectedProcesses returns the UniqueProcess information for all running
+	// processes that utilize the binaries previously specified by calls to
+	// AddPaths.
```

Comment:
Also note that it's expensive and that it's best to bundle everything into one call?

https://learn.microsoft.com/en-us/windows/win32/api/restartmanager/nf-restartmanager-rmregisterresources#remarks
> **Remarks**
> Each call to the RmRegisterResources function performs relatively expensive write operations. Do not call this function once per file, instead group related files together into components and register these together.

Reply:
Noted.